### PR TITLE
[EA] Onboarding Flow Redesign

### PR DESCRIPTION
**Pull request by oetherington**
This PR adds the redesigned onboarding flow for the EA forum as a followup to #8703 ([relevant Figma](https://www.figma.com/file/x7RwQ8kk2N5d4qaJZ3018O/Onboarding-2024?node-id=231%3A170&mode=dev)).

Screenshot 2024-02-12 at 18 18 33

Screenshot 2024-02-12 at 18 19 14

Screenshot 2024-02-12 at 18 19 26

Screenshot 2024-02-12 at 18 19 36

Screenshot 2024-02-12 at 18 19 50

Screenshot 2024-02-12 at 18 20 04

Screenshot 2024-02-12 at 18 20 12

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1206582728912027) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/components/ea-forum/auth/EALoginPopover.tsx [hunk 1]
```diff
@@ -282,15 +284,19 @@ export const EALoginPopover = ({open, setAction, isSignup, classes}: {
           ? client.signup(email, password)
           : client.login(email, password)
       );
-      window.location.reload();
+      await refetchCurrentUser();
```
Comment (oetherington) [inline packages/lesswrong/components/ea-forum/auth/EALoginPopover.tsx:287]: This allows us login/signup without refreshing the page

File: packages/lesswrong/components/ea-forum/onboarding/EAOnboardingFlow.tsx [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+  const {
+    BlurredBackgroundModal, EAOnboardingUserStage, EAOnboardingSubscribeStage,
+    EAOnboardingWorkStage, EAOnboardingThankYouStage,
+  } = Components;
+  return (
+    <BlurredBackgroundModal open className={classes.root}>
+      <EAOnboardingContextProvider onOnboardingComplete={onOnboardingComplete}>
+        <EAOnboardingUserStage />
```
Comment (oetherington) [inline packages/lesswrong/components/ea-forum/onboarding/EAOnboardingFlow.tsx:43]: We render _all_ the stages here (though all except one will return null). This allows us to do things like pre-loading the suggested tags and users to subscribe to before we get to that stage to make the flow a bit more seamless.

File: .eslintrc.js [hunk 3]
```diff
@@ -66,6 +66,7 @@ module.exports = {
     "react/no-unescaped-entities": 0,
     "react/display-name": 0,
     "react/jsx-no-comment-textnodes": 1,
+    "react/no-unknown-property": ["error", {ignore: ["test-id"]}],
```
Comment (oetherington) [inline .eslintrc.js:69]: I added a `test-id` to a component for easier access in cypress but it turns out that's disallowed by this lint rule and, IMO, it should be allowed.

**Comment by oetherington** [comment]
FYI @agnesstenlund

**Comment by agnesstenlund** [comment]
looks beautiful 🥲 

I'm doing some smaller UI polish on my side, one bigger thing I realised is that there is another screen I haven't accounted for, this one:
Screenshot 2024-02-12 at 18 59 33

it can probably stay the way it is, but I think ideally this step would look like the rest of the onboarding 🤔

**Comment by agnesstenlund** [comment]
the flow got interrupted a couple of the times I tried–I filled out the user name and then was suddenly in (skipping the next 3 screens). Not sure why and if it has to do with me updating the page etc. to see changes I've made locally 🤔

**Comment by agnesstenlund** [comment]
is there a way to check the onboarding flow without signing up with a new email? like a `http://localhost:3000/onboarding`

**Comment by agnesstenlund** [comment]
I think I may have messed up the selected state of the tags/ authors 🤔 I got thrown out of the flow again so am not sure

Ideally the selected state would look like this (sorry about not including anything in the designs 😅 )
Screenshot 2024-02-12 at 19 37 41
so darker image, background-colored border inside of the blue border, and a checkbox. Checkbox is not necessary though, I think people understand without it

**Comment by agnesstenlund** [comment]
a random thing I found–probably not related to this PR-is that I have 1 new notification after signing up. But then nothing showing on the notification page, and no new message in the inbox either 🤔

File: packages/lesswrong/components/Layout.tsx [hunk 4]
```diff
@@ -456,10 +457,11 @@ const Layout = ({currentUser, children, classes}: {
                     <FlashMessages />
                   </ErrorBoundary>
                   <ErrorBoundary>
-                    {showNewUserCompleteProfile
+                    {showNewUserCompleteProfile && !isEAForum
```
Comment (Stvad) [inline packages/lesswrong/components/Layout.tsx:460]: this can plausibly be FriendlyUI default? or is the content too EA specific 👀?

**Comment by jpaddison3** [comment]
> is that only showing on in dev

Correct

File: packages/lesswrong/components/Layout.tsx (hunk 4 above)
Comment (oetherington) [reply packages/lesswrong/components/Layout.tsx:460]: I'd guess it's a bit too specific - all the tags and users to subscribe to are hard-coded (at least for now), and the styles are meant to match the EA login popup which only supports Auth0. It wouldn't be much work to make it more generic in the future, but I don't think it's worth doing without a concrete use case.

**Comment by oetherington** [comment]
@agnesstenlund Thanks! Comments should be addressed now.

I don't think there's anything we can reasonably do to replace that permessions screen - it should only show when connecting a Google/Facebook account though (I think).

The page refreshes automatically when you make changes, and the flow won't be shown again if you already set your display name. There's not currently anyway to bring it back up.

I couldn't reproduce the notification bug, so I'll ignore for now but keep an eye out in the future.

**Review by s-cheng — APPROVED** [review]
This is really great, so excited to have it live! ASTLAC

One more small thing: I think we should update this error message to hint that the user might already have an account with that email address, since right now it feels like the user ran into a bug (when it's more likely that they already have an account)

Screen Shot 2024-02-12 at 12 46 20 PM

File: packages/lesswrong/components/ea-forum/onboarding/useEAOnboarding.tsx [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    await promise;
+    await goToNextStage();
+  }, [goToNextStage]);
+
+  if (!currentUser) {
+    const {LoginForm} = Components;
+    return (
+      <LoginForm />
+    );
+  }
```
[the remark this answers is not in the document]
Comment (oetherington) [reply packages/lesswrong/components/ea-forum/onboarding/useEAOnboarding.tsx:73]: Correct - I just thought this was better that throwing an exception. I'll add a comment.

**Comment by s-cheng** [comment]
Oh I forgot, can we include [this link](https://github.com/ForumMagnum/ForumMagnum/pull/8651) in the onboarding flow? @agnesstenlund FYI

(@oetherington I don't think this is blocking so feel free to merge in this PR without it)

**Pull request merged**